lib/varnishcfg/backends.go

```diff
@@ -0,0 +1,172 @@
+package varnishcfg
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/apache/trafficcontrol/lib/go-atscfg"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+)
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+func (v *VCLBuilder) configureDirectors(vclFile *vclFile, parents *atscfg.ParentAbstraction) ([]string, error) {
+	warnings := []string{}
+
+	vclFile.imports = append(vclFile.imports, "directors")
+	var err error
+	requestFQDNs := make([]string, 0)
+
+	for _, svc := range parents.Services {
+		addBackends(vclFile.backends, append(svc.Parents, svc.SecondaryParents...), svc.DestDomain, svc.Port)
+		addDirectors(vclFile.subroutines, svc)
+
+		requestFQDNs = []string{svc.DestDomain}
+
+		if v.toData.Server.Type == tc.CacheTypeEdge.String() {
+			dsRegexes := atscfg.MakeDSRegexMap(v.toData.DeliveryServiceRegexes)
+			anyCastPartners := atscfg.GetAnyCastPartners(v.toData.Server, v.toData.Servers)
+			requestFQDNs, err = atscfg.GetDSRequestFQDNs(
+				&svc.DS,
+				dsRegexes[tc.DeliveryServiceName(*svc.DS.XMLID)],
+				v.toData.Server,
+				anyCastPartners,
+				v.toData.CDN.DomainName,
+			)
+			if err != nil {
+				warnings = append(warnings, "error getting ds '"+*svc.DS.XMLID+"' request fqdns, skipping! Error: "+err.Error())
+				continue
+			}
+		}
+
+		assignBackends(vclFile.subroutines, svc, requestFQDNs)
+	}
+
+	return warnings, nil
+}
+
+func assignBackends(subroutines map[string][]string, svc *atscfg.ParentAbstractionService, requestFQDNs []string) {
+	lines := make([]string, 0)
+
+	conditions := make([]string, 0)
+	for _, fqdn := range requestFQDNs {
+		conditions = append(conditions, fmt.Sprintf(`req.http.host == "%s"`, fqdn))
+	}
+
+	lines = append(lines, fmt.Sprintf("if (%s) {", strings.Join(conditions, " || ")))
+	lines = append(lines, fmt.Sprintf("\tset req.backend_hint = %s.backend();", svc.Name))
+
+	// only change request host from edge servers which typically has multiple request FQDNs or
+	// one request FQDN that is not the origin.
+	if len(requestFQDNs) > 1 || (len(requestFQDNs) == 1 && requestFQDNs[0] != svc.DestDomain) {
+		lines = append(lines, fmt.Sprintf("\tset req.http.host = \"%s\";", svc.DestDomain))
```

Comment:
Might want to consider moving this to vcl_backend_fetch, and doing `set bereq.http.host=...` so that the host header is only changed on the request to the backend, and the original host header is preserved in the client request processing